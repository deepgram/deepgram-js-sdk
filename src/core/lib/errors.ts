export class DeepgramError extends Error {
  protected __dgError = true;

  constructor(message: string) {
    super(message);
    this.name = "DeepgramError";
  }
}

export function isDeepgramError(error: unknown): error is DeepgramError {
  return typeof error === "object" && error !== null && "__dgError" in error;
}

export class DeepgramApiError extends DeepgramError {
  status: number;

  constructor(message: string, status: number) {
    super(message);
    this.name = "DeepgramApiError";
    this.status = status;
  }

  toJSON() {
    return {
      name: this.name,
      message: this.message,
      status: this.status,
    };
  }
}

export class DeepgramUnknownError extends DeepgramError {
  originalError: unknown;

  constructor(message: string, originalError: unknown) {
    super(message);
    this.name = "DeepgramUnknownError";
    this.originalError = originalError;
  }
}

export class DeepgramVersionError extends DeepgramError {
  constructor() {
    super(
      `You are attempting to use an old format for a newer SDK version. Read more here: https://dpgr.am/js-v3`
    );

    this.name = "DeepgramVersionError";
  }
}

/**
 * Enhanced WebSocket error that captures additional debugging information
 * including status codes, request IDs, and response headers when available.
 */
export class DeepgramWebSocketError extends DeepgramError {
  originalEvent?: ErrorEvent | Event;
  statusCode?: number;
  requestId?: string;
  responseHeaders?: Record<string, string>;
  url?: string;
  readyState?: number;

  constructor(
    message: string,
    options: {
      originalEvent?: ErrorEvent | Event;
      statusCode?: number;
      requestId?: string;
      responseHeaders?: Record<string, string>;
      url?: string;
      readyState?: number;
    } = {}
  ) {
    super(message);
    this.name = "DeepgramWebSocketError";
    this.originalEvent = options.originalEvent;
    this.statusCode = options.statusCode;
    this.requestId = options.requestId;
    this.responseHeaders = options.responseHeaders;
    this.url = options.url;
    this.readyState = options.readyState;
  }

  toJSON() {
    return {
      name: this.name,
      message: this.message,
      statusCode: this.statusCode,
      requestId: this.requestId,
      responseHeaders: this.responseHeaders,
      url: this.url,
      readyState: this.readyState,
      originalEvent: this.originalEvent
        ? {
            type: this.originalEvent.type,
            timeStamp: this.originalEvent.timeStamp,
          }
        : undefined,
    };
  }
}
