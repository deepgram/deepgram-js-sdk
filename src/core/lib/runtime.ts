export const NODE_VERSION =
  typeof process !== "undefined" && process.versions && process.versions.node
    ? process.versions.node
    : "unknown";

export const BUN_VERSION =
  typeof process !== "undefined" && process.versions && process.versions.bun
    ? process.versions.bun
    : "unknown";

export const BROWSER_AGENT =
  typeof window !== "undefined" && window.navigator && window.navigator.userAgent
    ? window.navigator.userAgent
    : "unknown";

export const isBrowser = () => BROWSER_AGENT !== "unknown";

export const isNode = () => NODE_VERSION !== "unknown";

export const isBun = () => BUN_VERSION !== "unknown";
