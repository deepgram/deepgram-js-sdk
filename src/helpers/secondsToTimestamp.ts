export function secondsToTimestamp(seconds: number): string {
  return new Date(seconds * 1000).toISOString().substr(11, 12);
}
