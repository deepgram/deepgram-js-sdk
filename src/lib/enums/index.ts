export { LiveConnectionState } from "./LiveConnectionState";
export { LiveTranscriptionEvents } from "./LiveTranscriptionEvents";
