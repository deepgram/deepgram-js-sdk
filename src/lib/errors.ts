export class DeepgramError extends Error {
  protected __dgError = true;

  constructor(message: string) {
    super(message);
    this.name = "DeepgramError";
  }
}

export function isDeepgramError(error: unknown): error is DeepgramError {
  return typeof error === "object" && error !== null && "__dgError" in error;
}

export class DeepgramApiError extends DeepgramError {
  status: number;

  constructor(message: string, status: number) {
    super(message);
    this.name = "DeepgramApiError";
    this.status = status;
  }

  toJSON() {
    return {
      name: this.name,
      message: this.message,
      status: this.status,
    };
  }
}

export class DeepgramUnknownError extends DeepgramError {
  originalError: unknown;

  constructor(message: string, originalError: unknown) {
    super(message);
    this.name = "DeepgramUnknownError";
    this.originalError = originalError;
  }
}

export class DeepgramVersionError extends DeepgramError {
  constructor() {
    super(
      `You are attempting to use an old format for a newer SDK version. Read more here: https://dpgr.am/js-v3`
    );

    this.name = "DeepgramVersionError";
  }
}
