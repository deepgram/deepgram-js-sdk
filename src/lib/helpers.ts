import { Headers as CrossFetchHeaders } from "cross-fetch";
import {
  DeepgramClientOptions,
  FileSource,
  PrerecordedSource,
  UrlSource,
  TextSource,
  AnalyzeSource,
} from "./types";
import { Readable } from "stream";
import merge from "deepmerge";

export function stripTrailingSlash(url: string): string {
  return url.replace(/\/$/, "");
}

export const isBrowser = () => typeof window !== "undefined";
export const isServer = () => typeof process !== "undefined";

export function applySettingDefaults(
  options: DeepgramClientOptions,
  defaults: DeepgramClientOptions
): DeepgramClientOptions {
  return merge(defaults, options);
}

export function appendSearchParams(
  searchParams: URLSearchParams,
  options: Record<string, unknown>
): void {
  Object.keys(options).forEach((i) => {
    if (Array.isArray(options[i])) {
      const arrayParams = options[i] as Array<any>;
      arrayParams.forEach((param) => {
        searchParams.append(i, String(param));
      });
    } else {
      searchParams.append(i, String(options[i]));
    }
  });
}

export const resolveHeadersConstructor = () => {
  if (typeof Headers === "undefined") {
    return CrossFetchHeaders;
  }

  return Headers;
};

export const isUrlSource = (
  providedSource: PrerecordedSource | AnalyzeSource
): providedSource is UrlSource => {
  if ((providedSource as UrlSource).url) return true;

  return false;
};

export const isTextSource = (
  providedSource: PrerecordedSource | AnalyzeSource
): providedSource is TextSource => {
  if ((providedSource as TextSource).text) return true;

  return false;
};

export const isFileSource = (providedSource: PrerecordedSource): providedSource is FileSource => {
  if (isReadStreamSource(providedSource) || isBufferSource(providedSource)) return true;

  return false;
};

const isBufferSource = (providedSource: PrerecordedSource): providedSource is Buffer => {
  if (providedSource as Buffer) return true;

  return false;
};

const isReadStreamSource = (providedSource: PrerecordedSource): providedSource is Readable => {
  if (providedSource as Readable) return true;

  return false;
};

export class CallbackUrl extends URL {
  private callbackUrl = true;
}
