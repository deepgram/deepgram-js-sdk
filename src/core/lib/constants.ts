import { convertProtocolToWs } from "./helpers";
import { isBrowser, isBun, isNode, NODE_VERSION, BUN_VERSION, BROWSER_AGENT } from "./runtime";
import { version } from "./version";
import type { DefaultNamespaceOptions, DefaultClientOptions } from "./types";

const getAgent = () => {
  if (isNode()) {
    return `node/${NODE_VERSION}`;
  } else if (isBun()) {
    return `bun/${BUN_VERSION}`;
  } else if (isBrowser()) {
    return `javascript ${BROWSER_AGENT}`;
  } else {
    return `unknown`;
  }
};

export const DEFAULT_HEADERS = {
  "Content-Type": `application/json`,
  "X-Client-Info": `@deepgram/sdk; ${isBrowser() ? "browser" : "server"}; v${version}`,
  "User-Agent": `@deepgram/sdk/${version} ${getAgent()}`,
};

export const DEFAULT_URL = "https://api.deepgram.com";
export const DEFAULT_AGENT_URL = "wss://agent.deepgram.com";

export const DEFAULT_GLOBAL_OPTIONS: Partial<DefaultNamespaceOptions> = {
  fetch: { options: { url: DEFAULT_URL, headers: DEFAULT_HEADERS } },
  websocket: {
    options: { url: convertProtocolToWs(DEFAULT_URL), _nodeOnlyHeaders: DEFAULT_HEADERS },
  },
};

export const DEFAULT_AGENT_OPTIONS: Partial<DefaultNamespaceOptions> = {
  fetch: { options: { url: DEFAULT_URL, headers: DEFAULT_HEADERS } },
  websocket: {
    options: { url: DEFAULT_AGENT_URL, _nodeOnlyHeaders: DEFAULT_HEADERS },
  },
};

export const DEFAULT_OPTIONS: DefaultClientOptions = {
  global: DEFAULT_GLOBAL_OPTIONS,
  agent: DEFAULT_AGENT_OPTIONS,
};

export enum SOCKET_STATES {
  connecting = 0,
  open = 1,
  closing = 2,
  closed = 3,
}

export enum CONNECTION_STATE {
  Connecting = "connecting",
  Open = "open",
  Closing = "closing",
  Closed = "closed",
}
