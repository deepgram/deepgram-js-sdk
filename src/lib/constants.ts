import { isBrowser } from "./helpers";
import { FetchOptions } from "./types/Fetch";
import { version } from "./version";

export const DEFAULT_HEADERS = {
  "Content-Type": `application/json`,
  "X-Client-Info": `@deepgram/sdk; ${isBrowser() ? "browser" : "server"}; v${version}`,
  "User-Agent": `@deepgram/sdk/${version}`,
};

export const DEFAULT_URL = "https://api.deepgram.com";

export const DEFAULT_GLOBAL_OPTIONS = {
  url: DEFAULT_URL,
};

export const DEFAULT_FETCH_OPTIONS: FetchOptions = {
  headers: DEFAULT_HEADERS,
};

export const DEFAULT_OPTIONS = {
  global: DEFAULT_GLOBAL_OPTIONS,
  fetch: DEFAULT_FETCH_OPTIONS,
};
