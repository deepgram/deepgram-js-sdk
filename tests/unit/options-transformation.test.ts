import { convertLegacyOptions } from "../../src/lib/helpers";

describe("Unit Tests - Options Transformation Utilities", () => {
  describe("convertLegacyOptions", () => {
    it("should convert legacy options to new format", () => {
      const testCases = [
        {
          _experimentalCustomFetch: jest.fn(),
        },
        {
          restProxy: { url: "https://proxy.example.com" },
        },
        {
          global: { url: "https://custom-api.deepgram.com" },
        },
        {
          global: { headers: { "Custom-Header": "value" } },
        },
        {
          _experimentalCustomFetch: jest.fn(),
          restProxy: { url: "https://proxy.example.com" },
          global: { url: "https://custom-api.deepgram.com", headers: { Custom: "Header" } },
        },
        {},
      ];

      const results = testCases.map((options) => {
        const converted = convertLegacyOptions(options as any);
        // Replace function references with their type for snapshot testing
        return JSON.parse(
          JSON.stringify(converted, (key, value) => {
            if (typeof value === "function") {
              return "[Function]";
            }
            return value;
          })
        );
      });

      expect(results).toMatchSnapshot();
    });
  });
});
