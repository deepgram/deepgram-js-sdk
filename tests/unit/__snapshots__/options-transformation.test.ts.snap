// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit Tests - Options Transformation Utilities convertLegacyOptions should convert legacy options to new format 1`] = `
[
  {
    "_experimentalCustomFetch": "[Function]",
    "global": {
      "fetch": {
        "client": "[Function]",
      },
    },
  },
  {
    "global": {
      "fetch": {
        "options": {
          "proxy": {
            "url": "https://proxy.example.com",
          },
        },
      },
    },
    "restProxy": {
      "url": "https://proxy.example.com",
    },
  },
  {
    "global": {
      "fetch": {
        "options": {
          "url": "https://custom-api.deepgram.com",
        },
      },
      "url": "https://custom-api.deepgram.com",
      "websocket": {
        "options": {
          "url": "https://custom-api.deepgram.com",
        },
      },
    },
  },
  {
    "global": {
      "fetch": {
        "options": {
          "headers": {
            "Custom-Header": "value",
          },
        },
      },
      "headers": {
        "Custom-Header": "value",
      },
      "websocket": {
        "options": {
          "_nodeOnlyHeaders": {
            "Custom-Header": "value",
          },
        },
      },
    },
  },
  {
    "_experimentalCustomFetch": "[Function]",
    "global": {
      "fetch": {
        "client": "[Function]",
        "options": {
          "headers": {
            "Custom": "Header",
          },
          "proxy": {
            "url": "https://proxy.example.com",
          },
          "url": "https://custom-api.deepgram.com",
        },
      },
      "headers": {
        "Custom": "Header",
      },
      "url": "https://custom-api.deepgram.com",
      "websocket": {
        "options": {
          "_nodeOnlyHeaders": {
            "Custom": "Header",
          },
          "url": "https://custom-api.deepgram.com",
        },
      },
    },
    "restProxy": {
      "url": "https://proxy.example.com",
    },
  },
  {},
]
`;
